engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java

```diff
@@ -126,7 +126,7 @@ private Either<List<DirectBuffer>, CatchEventTuple> findErrorCatchEventInScope(
         if (errorCodeOptional.isPresent()) {
           final var errorCodeBuffer = errorCodeOptional.get();
           availableCatchEvents.getLeft().add(errorCodeBuffer);
-          if (errorCodeBuffer.equals(errorCode)) {
+          if (errorCodeBuffer.capacity() == 0 || errorCodeBuffer.equals(errorCode)) {
```

Comment:
:x: This doesn't introduce preference on code-specific Error catch events, so I would leave it as it is.

Instead, I would suggest the following:

1. We add an additional property to the class:
```java
private final CatchEventTuple catchAllEventTuple = new CatchEventTuple();
```
2. In the `findErrorCatchEventInScope()` method, we can have a separate check. This check shouldn't stop the search. It should only find the catch-all Error catch event at the lowest possible scope (since multiple catch-all Error events are possible at different scopes).
```java
  if (errorCodeBuffer.capacity() == 0 && catchAllEventTuple.instance != null) {
    catchAllEventTuple.instance = instance;
    catchAllEventTuple.catchEvent = catchEvent;
  }
```
3. In the `findErrorCatchEvent()`, before we create [the incident message](https://github.com/camunda/zeebe/blob/3bec81e620c46aa9998d10508cac77647fe5bfe6/engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java#L68), we can check if a catch-all Error events was found and return it.

This change would also allow us to introduce this behavior for Escalation events as well, allowing the use of code-specific Escalation events, and catch-all Escalation events. (I'll create a separate issue for this).

An alternative would be to use:

```java
Either<Map<DirectBuffer, CatchEventTuple>, CatchEventTuple> availableCatchEvents = Either.left(new HashMap<>());
```
to collect all the Error catch events we find. However, I can see two problems there:
1. We keep a lot of unnecessary data, since we only need the catch-all event.
2. The map will overwrite the value, if it find another catch-all event at a higher scope (unless we have a check for it).
3. It would require a larger refactoring for the Escalation code in this class.

Reply:
> In the findErrorCatchEvent(), before we create the incident message, we can check if a catch-all Error events was found and return it.

🤔 I think when an error is thrown, if we find that the code-specific error event is used in the current scope, if the current scope has a catch-all Error event, we should catch the error through the catch-all Error event, Instead of continuing to look for error events that catch specific code in the parent scope. I don't think we need to check again, see my comment above https://github.com/camunda/zeebe/pull/11141#discussion_r1048592619. WDYT?

Reply:
```
private final CatchEventTuple catchAllEventTuple = new CatchEventTuple();
```

As mentioned above, I don't think we need to introduce a new property. The following code can completely solve the current problem.

```

if (errorCodeBuffer.equals(errorCode)) {
    catchEventTuple.instance = instance;
    catchEventTuple.catchEvent = catchEvent;
    return Either.right(catchEventTuple);
}

// if the current scope has a catch-all Error event, we should catch the error through the catch-all Error event
if (errorCodeBuffer.capacity() == 0) {
    catchEventTuple.instance = instance;
    catchEventTuple.catchEvent = catchEvent;
    return Either.right(catchEventTuple);
}
```


 @koevskinikola. WDYT?

Reply:
:+1: Yes @lzgabel, you are correct! With what I suggested, I might cause the usage of a higher-scoped (code-specific) Error catch event, instead of a lower-scoped catch-all Error event.

In that case, I overengineered it. I agree, we don't need a new `catchAllEventTuple` property.

:thinking: But I don't think your suggestion would work for the following:

If the same scope has a catch-all Error event and a code-specific Error event, your code would always return the first one in the list.

If that is the catch-all Error event, it would be wrong, since we need to prefer the code-specific Error event.

If the list of Error catch events is first sorted so that the catch-all Error event comes last, it would work.

Reply:
Thanks a lot for pointing this out, I'll deal with it later. ❤️